### PR TITLE
 fix: skipping checking active volumes with the same number of files at the moment

### PR DESCRIPTION
**Pull request by kmlebedev**
# What problem are we solving?

 https://github.com/seaweedfs/seaweedfs/issues/4140

# How are we solving the problem?

If the volume modification time is close to the time of the last statistics collection pulse in the topology, then additionally check the number of vils at the moment for each such volume

# How is the PR tested?

localy
```
1. weed -v 4 server -s3 -filer -volume.max=0 -master.defaultReplication 001 -master.volumeSizeLimitMB=1024 -volume.preStopSeconds=1 -s3.port=8000 -s3.allowEmptyFolder=false -s3.allowDeleteBucketNotEmpty=false -s3.config=./docker/compose/s3.json
2. weed -v 4 volume -port 8081 -max=0 -preStopSeconds=1 -dir=/tmp
3. for i in {1..1000}; do s3cmd --access_key=some_access_key1 --secret_key=some_secret_key1 --no-ssl --host=127.0.0.1:8000 --host-bucket=http://127.0.0.1:8000 put ~/Downloads/any_doc.pdf "s3://test/any_doc.$i.pdf"; done
4. echo "volume.check.disk -v -force -syncDeleted" | weed shell
skipping active volumes 37 with the same file counts on 192.168.245.255:8080 and 192.168.245.255:8081
skipping active volumes 34 with the same file counts on 192.168.245.255:8080 and 192.168.245.255:8081
skipping active volumes 38 with the same file counts on 192.168.245.255:8080 and 192.168.245.255:8081
skipping active volumes 35 with the same file counts on 192.168.245.255:8080 and 192.168.245.255:8081
skipping active volumes 39 with the same file counts on 192.168.245.255:8080 and 192.168.245.255:8081
skipping active volumes 36 with the same file counts on 192.168.245.255:8080 and 192.168.245.255:8081
```

# Checks
- [ yes ] I have added unit tests if possible.
- [ no ] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

File: weed/shell/command_volume_check_disk.go [hunk 1]
```diff
@@ -7,16 +7,22 @@ import (
...
 	"google.golang.org/grpc"
 	"io"
 	"math"
 	"net/http"
+	"sync"
 	"time"
 )
 
+const (
+	VolumePulseSeconds = 5
```
Comment (chrislusf) [inline weed/shell/command_volume_check_disk.go]: Not a good place for this volume server setting.

File: weed/shell/command_volume_check_disk.go [hunk 2]
```diff
@@ -43,6 +49,66 @@ func (c *commandVolumeCheckDisk) Help() string {
...
+	go func() {
+		defer waitGroup.Done()
+		fileCountB, fileDeletedCountB = c.getVolumeStatusFileCount(b.info.Id, b.location.dataNode)
+	}()
+	// Trying to synchronize a remote call to two nodes
+	waitGroup.Wait()
+	return fileCountA == fileCountB, fileDeletedCountA == fileDeletedCountB
+}
+
+func (c *commandVolumeCheckDisk) doSkipVolume(a, b *VolumeReplica, timeBeforePulse int64, syncDeletions, verbose bool, writer io.Writer) bool {
```
Comment (chrislusf) [inline weed/shell/command_volume_check_disk.go]:
```suggestion
func (c *commandVolumeCheckDisk) shouldSkipVolume(a, b *VolumeReplica, timeBeforePulse int64, syncDeletions, verbose bool, writer io.Writer) bool {
```

File: weed/shell/command_volume_check_disk.go [hunk 3]
```diff
@@ -43,6 +49,66 @@ func (c *commandVolumeCheckDisk) Help() string {
...
+	doSyncDeletedCount := false
+	if syncDeletions && a.info.DeleteCount != b.info.DeleteCount {
+		doSyncDeletedCount = true
+	}
+	if a.info.FileCount != b.info.FileCount {
+		doSyncTotalFileCount = true
+	}
+	if doSyncTotalFileCount || doSyncDeletedCount {
+		if timeBeforePulse > a.info.ModifiedAtSecond && timeBeforePulse > b.info.ModifiedAtSecond {
+			return false
```
Comment (chrislusf) [inline weed/shell/command_volume_check_disk.go:93]: > If the volume modification time is close to the time of the last statistics collection pulse in the topology, then additionally check the number of vils at the moment for each such volume

Here the condition is the modification times are not close to stats collection time.

File: weed/shell/command_volume_check_disk.go [hunk 4]
```diff
@@ -43,6 +49,66 @@ func (c *commandVolumeCheckDisk) Help() string {
 `
 }
 
+func (c *commandVolumeCheckDisk) getVolumeStatusFileCount(vid uint32, dn *master_pb.DataNodeInfo) (fileCount, deletedFileCount uint64) {
+	_ = operation.WithVolumeServerClient(false, pb.NewServerAddressWithGrpcPort(dn.Id, int(dn.GrpcPort)), c.env.option.GrpcDialOption, func(volumeServerClient volume_server_pb.VolumeServerClient) error {
```
Comment (chrislusf) [inline weed/shell/command_volume_check_disk.go]: log the errors if any

File: weed/shell/command_volume_check_disk.go (hunk 2 above)
Comment (kmlebedev) [reply weed/shell/command_volume_check_disk.go]: ok

File: weed/shell/command_volume_check_disk.go (hunk 3 above)
Comment (kmlebedev) [reply weed/shell/command_volume_check_disk.go:93]: Yes, there is inverted logic.

Added a comment to the code for understanding.

File: weed/shell/command_volume_check_disk.go (hunk 4 above)
Comment (kmlebedev) [reply weed/shell/command_volume_check_disk.go]: ok

File: weed/shell/command_volume_check_disk.go (hunk 1 above)
Comment (kmlebedev) [reply weed/shell/command_volume_check_disk.go]: make new place

**Comment by kmlebedev** [comment]
@chrislusf corrected

**Pull request merged**